samples/snippets/src/main/java/product/ImportProductsBigQueryTable.java

```diff
@@ -0,0 +1,186 @@
+   */
+  private static final String DATASET_ID = "products";
+
+  /**
+   * This variable describes table id.
+   */
+  private static final String TABLE_ID = "products";
+
+  /**
+   * This variable describes data schema.
+   */
+  private static final String DATA_SCHEMA = "product";
+
+  /* TO CHECK ERROR HANDLING USE THE TABLE WITH INVALID PRODUCTS:
+     TABLE_ID = "products_some_invalid"
+  */
+
+  private ImportProductsBigQueryTable() {
+  }
+
+  /**
+   * Get product service client.
+   *
+   * @return ProductServiceClient.
+   * @throws IOException if endpoint is incorrect.
+   */
+  private static ProductServiceClient getProductServiceClient()
+      throws IOException {
+    return ProductServiceClient.create();
+  }
+
+  /**
+   * Get import products from big query request.
+   *
+   * @param reconciliationMode enum variable describes reconciliation.
+   * @return ImportProductsRequest.
+   */
+  public static ImportProductsRequest getImportProductsBigQueryRequest(
+      final ReconciliationMode reconciliationMode) {
+      /* TO CHECK ERROR HANDLING PASTE THE INVALID CATALOG NAME HERE:
+         DEFAULT_CATALOG = "invalid_catalog_name"
+      */
+    BigQuerySource bigQuerySource = BigQuerySource.newBuilder()
+        .setProjectId(PROJECT_ID)
+        .setDatasetId(DATASET_ID)
+        .setTableId(TABLE_ID)
+        .setDataSchema(DATA_SCHEMA)
+        .build();
+
+    ProductInputConfig inputConfig = ProductInputConfig.newBuilder()
+        .setBigQuerySource(bigQuerySource)
+        .build();
+
+    ImportProductsRequest importRequest = ImportProductsRequest.newBuilder()
+        .setParent(DEFAULT_CATALOG)
+        .setReconciliationModeValue(reconciliationMode.getNumber())
+        .setInputConfig(inputConfig)
+        .build();
+
+    System.out.printf("Import products from big query table request: %s%n",
+        importRequest);
+
+    return importRequest;
+  }
+
+  /**
+   * Call the Retail API to import products.
+   *
+   * @throws IOException          from the called method.
+   * @throws ExecutionException   when attempting to retrieve the result of a
+   *                              task that aborted by throwing an exception.
+   * @throws InterruptedException when a thread is waiting, sleeping, or
+   *                              otherwise occupied, and the thread is
+   *                              interrupted, either before or during the
+   *                              activity.
+   */
+  public static void importProductsFromBigQuery()
+      throws IOException, ExecutionException, InterruptedException {
+    // TRY THE FULL RECONCILIATION MODE HERE:
+    ReconciliationMode reconciliationMode = ReconciliationMode.INCREMENTAL;
+
+    ImportProductsRequest importBigQueryRequest =
+        getImportProductsBigQueryRequest(reconciliationMode);
+
+    OperationFuture<ImportProductsResponse, ImportMetadata> bigQueryOperation =
+        getProductServiceClient().importProductsAsync(importBigQueryRequest);
```

Comment:
One caveat for using operationFuture is that this is returning a Java future which will expire in 5 minutes, after which future.get() will return a CancelledException, and product import usually will normally take more than 5 minutes. We can work around this by pulling the Operation manually. Example for this new style:

        ProductServiceClient serviceClient = getProductServiceClient();
        String operationName = serviceClient
                .importProductsCallable()
                .call(importBigQueryRequest)
                .getName();
        System.out.printf("OperationName = %s", operationName);
        OperationsClient operationsClient = serviceClient.getOperationsClient();
        Operation operation = operationsClient.getOperation(operationName);
        while (!operation.getDone()) {
            // Keep polling the operation periodically until the import task is done.
            Thread.sleep(60000);
            operation = operationsClient.getOperation(operationName);
        }
        if (operation.hasMetadata()) {
            ImportMetadata metadata = operation.getMetadata().unpack(ImportMetadata.class);
            System.out.printf("Number of successfully imported products: %s\n",
                    metadata.getSuccessCount());
            System.out.printf("Number of failures during the importing: %s\n",
                    metadata.getFailureCount());
        }
        if (operation.hasResponse()) {
            ImportProductsResponse response = operation.getResponse().unpack(ImportProductsResponse.class);
            System.out.printf("Number of successfully imported products: %s\n", response);
        }

Reply:
done